### PR TITLE
ctpdev: scalers/config dev

### PR DESCRIPTION
**Pull request by lietava**

### CODE REVIEW

**Comment by lietava** [comment]
@shahor02 : is there a way to flush logs ?

**Comment by shahor02** [comment]
@lietava what logs you refer to? If you ask why you don't see CI reports, this is because the CI does not stat for draft PRs.

**Comment by lietava** [comment]
@shahor02 sorry, I was not clear. I mean logs from code, i.e. ctp-proxy produced by LOG. Does std::flush help ?

**Comment by shahor02** [comment]
The LOG is flushed instantly, you don't need to do anything.

**Comment by lietava** [comment]
But it seems to me that I see difference between running directly from command line
and nohup ... > my.log - when I tail -f my.log ?
I mean tail is not showing up to the end of actual output ?

**Comment by shahor02** [comment]
Actually, I also noticed this when running with `| tee ...` but I thought there is some bufferization on the tee level. 
At least I remember the Logger documentation mentioning that it is flushing instantly. Perhaps @ktf may comment?

**Comment by lietava** [comment]
@shahor02 : different question: when I start o2-ctp-proxy I see four processes with name o2-ctp-proxy running. Is it ok ?

**Comment by shahor02** [comment]
could you shot the `top` screenshot (with `c` to see full name)?

**Comment by lietava** [comment]
image

**Comment by lietava** [comment]
IN top number of o2 changes from 1 to 4

**Comment by shahor02** [comment]
There are 3 real devices shown by top, corresponding to these devices:
```
o2-ctp-proxy -h
...
Available data processors (full info with '--help full'):
  --internal-dpl-clock arg               No processor options
  --ctp-proxy arg                        Use '--help ctp-proxy' to display 
                                         processor options
  --internal-dpl-injected-dummy-sink arg No processor options
```
The `pgrep -a o2-` shows 3 of them + simply `o2-ctp-proxy`, apparently, this is a process which spawns others.
Anyway, nothing anomalous.

**Comment by lietava** [comment]
Thanks.

**Comment by lietava** [comment]
@shahor02 can I see browse CCDB objects in browser ?

**Comment by shahor02** [comment]
http://ccdb-test.cern.ch:8080/browse/?report=true or http://alice-ccdb.cern.ch/browse 
See https://github.com/AliceO2Group/AliceO2/blob/dev/CCDB/README.md#access-with-a-browser

**Comment by lietava** [comment]
@shahor02 it is waiting for checks  for 4 days. Is it ok ?

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h [hunk 1]
```diff
@@ -111,7 +111,7 @@ class CTPConfiguration
                     CLUSTER,
                     CLASS,
                     UNKNOWN };
-  int loadConfigurationRun3(const std::string& ctpconfiguartion);
+  int loadConfigurationRun3(const std::string ctpconfiguartion);
```
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: why? using the reference was correct thing to do.

File: DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h [hunk 2]
```diff
@@ -162,8 +163,8 @@ class CTPRunManager
   void init();
   int startRun(std::string& cfg);
   int stopRun(uint32_t irun);
-  int addScalers(uint32_t irun);
-  int processMessage(std::string& message);
+  int addScalers(uint32_t irun, std::time_t time);
+  int processMessage(std::string& topic, std::string& message);
```
Comment (shahor02) [inline DataFormats/Detectors/CTP/include/DataFormatsCTP/Configuration.h]: I think both strings can be `const string&`

**Review by shahor02 — APPROVED** [review]
Hi @lietava 
please see 2 cosmetic comments below, I will merge it as is, but please fix this in the next PR

**Review by shahor02 — CHANGES_REQUESTED** [review]
wait, you did not remove the draft tag and that's why it does not start the tests.

**Comment by lietava** [comment]
@shahor02 : fixed.

**Comment by lietava** [comment]
It failed in build/O2/fullCI. Sorry, I can not find what is the reason . Can you, let me know, please ?

**Comment by shahor02** [comment]
@lietava this is an old failure from the already fixed bug. Will add dummy commit to retrigger CI

**Comment by lietava** [comment]
@shahor02 it failed.

**Comment by shahor02** [comment]
It fails trying to read your object from the CCDB:

```
./sim-challenge.log[21315:CTPDigitizer]: Info in <TJAlienFile::Open>: Accessing file /alice/data/CCDB/CTP/Config/Config/15/19450/88f90b32-cbc4-11ec-b790-200114580202 in SE <ALICE::CERN::EOS>
./sim-challenge.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class string read too few bytes: 2 instead of 8
./sim-challenge.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class vector<o2::ctp::CTPGenerator> read too few bytes: 121 instead of 229
./sim-challenge.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class string read too few bytes: 76 instead of 486
./sim-challenge.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class vector<o2::ctp::CTPInput> read too many bytes: 712 instead of 580
./sim-challenge.log[21315:CTPDigitizer]: Warning in <TBufferFile::CheckByteCount>: vector<o2::ctp::CTPInput>::Streamer() not in sync with data on file alien:///alice/data/CCDB/CTP/Config/Config/15/19450/88f90b32-cbc4-11ec-b790-200114580202, fix Streamer()
./sim-challenge.log[21315:CTPDigitizer]: Error in <TBufferFile::ReadVersion>: Could not find the StreamerInfo with a checksum of 0x40000006 for the class "string" in alien:///alice/data/CCDB/CTP/Config/Config/15/19450/88f90b32-cbc4-11ec-b790-200114580202.
./sim-challenge.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class string read too few bytes: 7 instead of 206307328
./sim-challenge.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: Byte count probably corrupted around buffer position 975:
./sim-challenge.log[21315:CTPDigitizer]: 	206307328 for a possible maximum of 556

./digi.log[21315:CTPDigitizer]: Info in <TJAlienFile::Open>: Accessing file /alice/data/CCDB/CTP/Config/Config/15/19450/88f90b32-cbc4-11ec-b790-200114580202 in SE <ALICE::CERN::EOS>
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class string read too few bytes: 2 instead of 8
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class vector<o2::ctp::CTPGenerator> read too few bytes: 121 instead of 229
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class string read too few bytes: 76 instead of 486
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class vector<o2::ctp::CTPInput> read too many bytes: 712 instead of 580
./digi.log[21315:CTPDigitizer]: Warning in <TBufferFile::CheckByteCount>: vector<o2::ctp::CTPInput>::Streamer() not in sync with data on file alien:///alice/data/CCDB/CTP/Config/Config/15/19450/88f90b32-cbc4-11ec-b790-200114580202, fix Streamer()
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::ReadVersion>: Could not find the StreamerInfo with a checksum of 0x40000006 for the class "string" in alien:///alice/data/CCDB/CTP/Config/Config/15/19450/88f90b32-cbc4-11ec-b790-200114580202.
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class string read too few bytes: 7 instead of 206307328
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: Byte count probably corrupted around buffer position 975:
./digi.log[21315:CTPDigitizer]: 	206307328 for a possible maximum of 556
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class vector<o2::ctp::CTPDetector> read too many bytes: 587 instead of 32
./digi.log[21315:CTPDigitizer]: Warning in <TBufferFile::CheckByteCount>: vector<o2::ctp::CTPDetector>::Streamer() not in sync with data on file alien:///alice/data/CCDB/CTP/Config/Config/15/19450/88f90b32-cbc4-11ec-b790-200114580202, fix Streamer()
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::ReadClassBuffer>: Could not find the StreamerInfo for version 3 of the class o2::gpu::gpustd::bitset<32>, object skipped at offset 1085
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::ReadClassBuffer>: Could not find the StreamerInfo for version 4 of the class o2::gpu::gpustd::bitset<32>, object skipped at offset 1093
./digi.log[21315:CTPDigitizer]: Error in <TBufferFile::CheckByteCount>: object of class vector<o2::ctp::CTPCluster> read too few bytes: 100 instead of 547
```
Looks like you have modified your classes in a way the automatic schema evolution cannot support it. 
Which classes were changed? You should probably add manual schema evolution (https://root.cern/manual/io/#manual-schema-evolution-user-defined-io-customization-rules) for these class to be backward compatible and avoid painful transition period and create / upload new object.

**Comment by lietava** [comment]
@shahor02 can it be I changed Configuration.h data and did not increase version ?
Ohh, I see now your comment. Yes, I did.

**Comment by lietava** [comment]
@shahor02  Is it not enough if I increase classdef which I did not ?

**Comment by shahor02** [comment]
You should do this in anyway, let's see if this fixes the problem. But, usually the error due to the wrong class version is different.

**Comment by lietava** [comment]
@shahor02 I know I should have done it, sorry. Other question so far there are now ctp config data.
This check can not be reset from now ?

**Comment by shahor02** [comment]
> @shahor02 I know I should have done it, sorry. Other question so far there are now ctp config data. This check can not be reset from now ?

Sorry, not sure I understood what you mean. There is an object currently loaded to the CCDB and it is read by e.g. Digitization. If you override it with the new object in the format, all tests with O2 not including this PR will fail. Let's see if simply fixing the class version helps.

**Review by shahor02 — APPROVED** [review]
ok, in the end it was a class version problem...

**Pull request merged**